Add deprecation check for Java version

pugnascotia opened the pull request:
Supersedes PR #41092.

Adds a deprecation check for Java version, as Java 11 will be required
in the next version of Elasticsearch.

Relates to #40756 and #40754.

AthenaEryma committed:
Add deprecation check for Java version

Adds a deprecation check for Java version, as Java 11 will be required
in the next version of Elasticsearch.

pugnascotia committed:
Merge remote-tracking branch 'upstream/7.x' into deprecation-check-java

pugnascotia committed:
Merge remote-tracking branch 'upstream/7.x' into deprecation-check-java

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Core)

pugnascotia committed:
Test fixes across JVM versions

pugnascotia committed:
More fixes

rjernst reviewed (approved):
Looks good. One minor question: it looks like this will only exist in the Kibana migration UI, not in the deprecation log. Should we also have it in the latter?

pugnascotia commented:
@rjernst this was already being logged in the `Bootstrap` class.

pugnascotia committed:
Merge remote-tracking branch 'upstream/7.x' into deprecation-check-java

pugnascotia committed:
Disable node deprecation length checks in REST tests

Some tests check the length of the `node_settings` array that comes back
from the `/_migration/deprecations` API. However, the length can now
vary depending on the JDK being used by ES, and that can't be
represented in the REST tests at the moment.

pugnascotia commented on x-pack/plugin/src/test/resources/rest-api-spec/test/deprecation/10_basic.yml:
```diff
@@ -10,7 +10,10 @@ setup:
+  # We don't check this field because for ES <8.0 && Java < 11, a node-level deprecation warning is triggered
+  # and it isn't easy to codify this in the REST tests without extra test infrastructure. `MigrationIT.java`
+  # does check this, however.
+  # - length: { node_settings: 0 }
```
@rjernst can you can think of a better approach here?

rjernst replied:
I think this is fine. Thanks for the note point to MigrationIT

rjernst commented on x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/NodeDeprecationChecks.java:
```diff
@@ -255,4 +256,15 @@ static DeprecationIssue checkRemovedSetting(final Settings settings, final Setti
+            return new DeprecationIssue(DeprecationIssue.Level.CRITICAL,
+                "Java 11 is required",
+                "https://www.elastic.co/guide/en/elasticsearch/reference/master/breaking-changes-8.0.html#_java_11_is_required",
+                "Java 11 will be required for future versions of Elasticsearch, this node is running version "
```
I wonder if we should first suggest using the bundled jdk?

pugnascotia replied:
Good idea. I've added a note here, and in a couple other places too. What do you think?

rjernst replied:
The new wording is better, but it may be confusing for some users since they may have the bundled jdk, but have JAVA_HOME set by their system, so what they need to do is unset JAVA_HOME.

pugnascotia replied:
I've expanded the message further.

rjernst reviewed (approved):
One more minor comment, still looks good.

pugnascotia committed:
Reference bundled JDK in deprecation message

pugnascotia committed:
Reference bundled JDK in other deprecation messages also

pugnascotia committed:
Expand JDK message further

rjernst reviewed (approved):
LGTM

pugnascotia commented:
Thanks to @gwbrown for the initial implementation.
